File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        
        int n = nums.length;
        int pref[] = new int[n];
        pref[0] = nums[0];
        for(int i = 1;i<n;i++){
            pref[i] = pref[i-1] + nums[i];
        }
        int count = 0;
        HashMap<Integer,Integer> map = new HashMap<>();
        map.put(0,1);
        for(int i = 0;i<n;i++){
            int diff = pref[i] - k;
            if(map.containsKey(diff)){
                count+=map.get(diff);
            }
            map.put(pref[i],map.getOrDefault(pref[i],0)+1);
        }
        return count;
    }
}

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
//check the hint
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n=nums.size();
        vector<int>prefix_sum(n);
        prefix_sum[0]=nums[0];
        for(int i=1;i<n;i++){
            prefix_sum[i]=prefix_sum[i-1]+nums[i];
        }
        int res=0;
        unordered_map<int,int>mp;
        //corner case
        mp[0]=1;
        for(int i=0;i<n;i++){
            int val=prefix_sum[i]-k;
            res+=mp[val];
            mp[prefix_sum[i]]++;
        }
        return res;
    }
};